73-offline-attestation.md

```diff
@@ -0,0 +1,369 @@
+nitty-gritty.
+-->
+
+- The `registrar` will be modified to, upon initial `agent` registration -
+  which includes the execution of `tpm2_makecredential` - record the EK, AIK
+into a JSON file, sign it (using the private key generated as part of the
+certificates for mTLS interaction with both `tenant` and `registrar`) and then
+make a record of it on a tamper-resistant transparency log (e.g., Rekor). In
+addition to that, it will store the JSON file, the signature, and the public
+key on the time-series like persistent datastore. This should allow an external
+component/user to check, provided that there is trust on the `registrar`, that
+a particular AIK is indeed tied to a particular EK. The reason for having this
+data stored into a time-series is due to the fact that AIKs are regenereated
+every time an `agent` is restarted on Keylime.
+- The `verifier` will be modified to take the `json_response` (python
+  dictionary) from the `agent` - which will include both TPM quotes and logs (MB
+and IMA) - plus `agent` data (python dictionary which includes all columns stored)
+from the SQL database, internal to Keylime), combine it into a single (python 
+dictionary record) and store it on a time-series datastore.
+- The `verifier` will also be modified to extract the "TPM clock information"
+ (i.e, "clock", "resetCount", "restartCount", "safe") from the quote 
+(part of `json_response`) and make it available on this same python
+dictionary. The `verifier` will then use this information to compare this
+ timestamp with the "previous" one (in the case of "online" Keylime 
+attestation, the one stored on the database, as part of the `agent` data
+ python dictionary) in order to decide if a quote is being replayed. This 
+mechanism is not strictly necessary for the online attestation (given trust
+in the `verifier`), but will become crucial for offline attestation, where 
+entries on the time-series like persistent datastore could be modified 
+by an attacker. 
+- Three new parameters: the name of a python module to be dynamically
+  imported (which will contain code used to interact with these new proposed
+stores) and the URLs for these these new proposed stores will be
+supplied by the user as parameters under the `[general]` section: 
+`durable_attestation_import`, `persistent_store_url` and 
+`transparency_log_url`. The URL format is similar to the one already
+used to establish connectivity to SQL databases within keylime.
+- First additional tunable: an user can specify, via an attribute on 
+`keylime.conf` (`[cloud_verifier]` section), additional attributes of the
+ `agent` python dictionary to be signed by the `verifier`, being this 
+signing also stored on tamper-resistant transparency log. The value of 
+this attribute - `signed_attributes` - could be set to `all`, which will 
+result in the `verifier` signing **everything**, including the 
+`json_response` data package from the `agent`, and storing a record on the 
```

Comment:
The signature of the contents of every single piece of data required by attestation would imply trust in the `verifier`, something that we are trying to avoid. Making it optional does not prevent a deployment in a scenario where this (trust in the `verifier`) is the case, while still accommodating "our" (IBM cloud) production scenario.